### PR TITLE
minor: add a diagram to docstring for DictionaryArray

### PR DESCRIPTION
**Pull request by alamb**
Rationale:

I made a diagram while working with @AssHero on https://github.com/apache/arrow-datafusion/pull/2721 which I felt would be generally helpful as the `keys` and `values` terminology often gets confusing.

Changes:
Add some more docstrings to `DictionaryArray` based on  https://github.com/apache/arrow-datafusion/pull/2721#discussion_r901091846

### CODE REVIEW

**Review by HaoYang670 — APPROVED** [review]
LGTM! Just leave a nit.
`DictionaryArray` is always used when you have data with many repeated values. So maybe it is better to give an example with short and fat `values` array  and long `keys` array.

File: arrow/src/array/array_dictionary.rs [hunk 1]
```diff
@@ -33,6 +33,31 @@ use crate::error::Result;
 /// This is mostly used to represent strings or a limited set of primitive types as integers,
 /// for example when doing NLP analysis or representing chromosomes by name.
 ///
+/// [`DictionaryArray`] are represented using a `keys` array and a
+/// `values` array, which may be diferent lengths. The `keys` array
+/// stores indexes in the `values` array which holding holds
```
Comment (jhorstmann) [inline arrow/src/array/array_dictionary.rs:38]:
```suggestion
/// stores indexes in the `values` array which holds
```

**Review by jhorstmann — APPROVED** [review]

File: arrow/src/array/array_dictionary.rs [hunk 2]
```diff
@@ -33,6 +33,31 @@ use crate::error::Result;
 /// This is mostly used to represent strings or a limited set of primitive types as integers,
 /// for example when doing NLP analysis or representing chromosomes by name.
 ///
+/// [`DictionaryArray`] are represented using a `keys` array and a
+/// `values` array, which may be diferent lengths. The `keys` array
```
Comment (waynexia) [inline arrow/src/array/array_dictionary.rs:37]:
```suggestion
/// `values` array, which may be different lengths. The `keys` array
```

**Review by waynexia — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by alamb** [comment]
> DictionaryArray is always used when you have data with many repeated values. So maybe it is better to give an example with short and fat values array and long keys array.

Good point @HaoYang670  -- I updated the diagram in https://github.com/apache/arrow-rs/pull/1909/commits/8cf0dd3d68d6c7ec0f2d41574b0ac13fa7b20a8e -- what do you think?

**Pull request merged**